File: chroma-league/src/test/java/net/booone/chromaleague/rest/eventdata/EventTypeTest.java
package net.booone.chromaleague.rest.eventdata;

import net.booone.chromaleague.GameStateMocks;
import net.booone.chromaleague.ResourceLoader;
import net.booone.chromaleague.hud.parts.dragons.KilledDragonsBarTest;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;

import java.util.stream.Stream;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.when;

class EventTypeTest {
    private GameStateMocks gameStateMocks;

    @BeforeEach
    void setUp() {
        gameStateMocks = new GameStateMocks();
    }

    @ParameterizedTest(name = "{0} -> {1}")
    @MethodSource("dragonToExpectedTypeArguments")
    void testDragonsAnimation(DragonType dragonType, EventType expectedEventType, boolean alliedKill) {
        //given
        when(gameStateMocks.playerList().isAlly(any())).thenReturn(alliedKill);
        final Event event = KilledDragonsBarTest.getEventForDragonKill(dragonType);

        //when
        final EventType eventType = EventType.fromEvent(event);

        //then
        assertEquals(expectedEventType, eventType);
    }

    static Stream<Arguments> dragonToExpectedTypeArguments() {
        return Stream.of(
                Arguments.of(DragonType.OCEAN, EventType.ALLY_OCEAN_DRAGON_KILL, true),
                Arguments.of(DragonType.MOUNTAIN, EventType.ALLY_MOUNTAIN_DRAGON_KILL, true),
                Arguments.of(DragonType.INFERNAL, EventType.ALLY_INFERNAL_DRAGON_KILL, true),
                Arguments.of(DragonType.CLOUD, EventType.ALLY_CLOUD_DRAGON_KILL, true),
                Arguments.of(DragonType.CHEMTECH, EventType.ALLY_CHEMTECH_DRAGON_KILL, true),
                Arguments.of(DragonType.HEXTECH, EventType.ALLY_HEXTECH_DRAGON_KILL, true),
                Arguments.of(DragonType.ELDER, EventType.ALLY_ELDER_DRAGON_KILL, true),
                Arguments.of(DragonType.OCEAN, EventType.ENEMY_OCEAN_DRAGON_KILL, false),
                Arguments.of(DragonType.MOUNTAIN, EventType.ENEMY_MOUNTAIN_DRAGON_KILL, false),
                Arguments.of(DragonType.INFERNAL, EventType.ENEMY_INFERNAL_DRAGON_KILL, false),
                Arguments.of(DragonType.CLOUD, EventType.ENEMY_CLOUD_DRAGON_KILL, false),
                Arguments.of(DragonType.CHEMTECH, EventType.ENEMY_CHEMTECH_DRAGON_KILL, false),
                Arguments.of(DragonType.HEXTECH, EventType.ENEMY_HEXTECH_DRAGON_KILL, false),
                Arguments.of(DragonType.ELDER, EventType.ENEMY_ELDER_DRAGON_KILL, false)
        );
    }

    @Test
    void testActivePlayerKill() {
        //given
        when(gameStateMocks.playerList().isAlly(any())).thenReturn(true);
        final Event event = ResourceLoader.eventFromJson("activePlayerKill.json");

        //when
        final EventType eventType = EventType.fromEvent(event);

        //then
        assertEquals(EventType.ACTIVE_PLAYER_KILL, eventType);
    }

    @Test
    void testActivePlayerAssist() {
        //given
        when(gameStateMocks.playerList().isAlly(any())).thenReturn(true);
        final Event event = ResourceLoader.eventFromJson("activePlayerAssist.json");

        //when
        final EventType eventType = EventType.fromEvent(event);

        //then
        assertEquals(EventType.ACTIVE_PLAYER_ASSIST, eventType);
    }

    @Test
    void testAllyHeraldKillEvent() {
        //given
        when(gameStateMocks.playerList().isAlly(any())).thenReturn(true);
        final Event event = ResourceLoader.eventFromJson("allyHeraldKill.json");

        //when
        final EventType eventType = EventType.fromEvent(event);

        //then
        assertEquals(EventType.ALLY_HERALD_KILL, eventType);
    }

    @Test
    void testEnemyHeraldKillEvent() {
        //given
        when(gameStateMocks.playerList().isAlly(any())).thenReturn(false);
        final Event event = ResourceLoader.eventFromJson("allyHeraldKill.json");

        //when
        final EventType eventType = EventType.fromEvent(event);

        //then
        assertEquals(EventType.ENEMY_HERALD_KILL, eventType);
    }

    @Test
    void testAllyBaronKillEvent() {
        //given
        when(gameStateMocks.playerList().isAlly(any())).thenReturn(true);
        final Event event = ResourceLoader.eventFromJson("allyBaronKill.json");

        //when
        final EventType eventType = EventType.fromEvent(event);

        //then
        assertEquals(EventType.ALLY_BARON_KILL, eventType);
    }

    @Test
    void testEnemyBaronKillEvent() {
        //given
        when(gameStateMocks.playerList().isAlly(any())).thenReturn(false);
        final Event event = ResourceLoader.eventFromJson("allyBaronKill.json");

        //when
        final EventType eventType = EventType.fromEvent(event);

        //then
        assertEquals(EventType.ENEMY_BARON_KILL, eventType);
    }

    @Test
    void testGameEndWinEvent() {
        //given
        final Event event = ResourceLoader.eventFromJson("gameEndWin.json");

        //when
        final EventType eventType = EventType.fromEvent(event);

        //then
        assertEquals(EventType.GAME_END_VICTORY, eventType);
    }

    @Test
    void testGameEndLoseEvent() {
        //given
        final Event event = ResourceLoader.eventFromJson("gameEndLose.json");

        //when
        final EventType eventType = EventType.fromEvent(event);

        //then
        assertEquals(EventType.GAME_END_DEFEAT, eventType);
    }
}

File: test/rest/eventdata/EventType.test.ts
import {beforeEach, describe, expect, test} from 'vitest';
import {EventType, eventTypeFromEvent} from '../../../src/rest/eventdata/EventType.js';
import {DragonType} from '../../../src/rest/eventdata/DragonType.js';
import type {Event} from '../../../src/rest/eventdata/Event.js';
import {RunningState} from '../../../src/state/RunningState.js';
import {PlayerList} from '../../../src/rest/playerlist/PlayerList.js';
import type {Player} from '../../../src/rest/playerlist/Player.js';
import {Team} from '../../../src/rest/playerlist/Team.js';

const PLAYER_RIOT_ID = 'BonE#EUPL';
const PLAYER_RIOT_ID_GAME_NAME = 'BonE';

function setupGameState(isAlly: boolean): void {
    RunningState.setRunningGame(false);
    RunningState.setRunningGame(true);
    RunningState.getGameState().playerRiotId = PLAYER_RIOT_ID;
    RunningState.getGameState().activePlayer = {
        riotId: PLAYER_RIOT_ID,
        riotIdGameName: PLAYER_RIOT_ID_GAME_NAME,
        currentGold: 0,
        level: 1,
        championStats: { attackRange: 125, currentHealth: 100, maxHealth: 100, resourceMax: 100, resourceValue: 100, resourceType: 'MANA' },
    };

    const allyPlayers: Player[] = [
        { riotId: PLAYER_RIOT_ID, riotIdGameName: PLAYER_RIOT_ID_GAME_NAME, team: Team.CHAOS, championName: "Cho'Gath", respawnTimer: 0, isDead: false },
        { riotId: 'BooonE#TEST', riotIdGameName: 'BooonE', team: isAlly ? Team.CHAOS : Team.ORDER, championName: 'Test', respawnTimer: 0, isDead: false },
    ];
    RunningState.getGameState().playerList = new PlayerList(allyPlayers);
}

function dragonEvent(dragonApiType: string): Event {
    return { EventID: 7, EventName: 'DragonKill', EventTime: 933, DragonType: dragonApiType, KillerName: 'BooonE' };
}

describe('EventTypeTest', () => {
    beforeEach(() => {
        RunningState.setRunningGame(false);
        RunningState.setRunningGame(true);
    });

    const dragonToExpectedTypeArguments: [DragonType, string, EventType, boolean][] = [
        [DragonType.OCEAN, 'Water', EventType.ALLY_OCEAN_DRAGON_KILL, true],
        [DragonType.MOUNTAIN, 'Earth', EventType.ALLY_MOUNTAIN_DRAGON_KILL, true],
        [DragonType.INFERNAL, 'Fire', EventType.ALLY_INFERNAL_DRAGON_KILL, true],
        [DragonType.CLOUD, 'Air', EventType.ALLY_CLOUD_DRAGON_KILL, true],
        [DragonType.CHEMTECH, 'Chemtech', EventType.ALLY_CHEMTECH_DRAGON_KILL, true],
        [DragonType.HEXTECH, 'Hextech', EventType.ALLY_HEXTECH_DRAGON_KILL, true],
        [DragonType.ELDER, 'Elder', EventType.ALLY_ELDER_DRAGON_KILL, true],
        [DragonType.OCEAN, 'Water', EventType.ENEMY_OCEAN_DRAGON_KILL, false],
        [DragonType.MOUNTAIN, 'Earth', EventType.ENEMY_MOUNTAIN_DRAGON_KILL, false],
        [DragonType.INFERNAL, 'Fire', EventType.ENEMY_INFERNAL_DRAGON_KILL, false],
        [DragonType.CLOUD, 'Air', EventType.ENEMY_CLOUD_DRAGON_KILL, false],
        [DragonType.CHEMTECH, 'Chemtech', EventType.ENEMY_CHEMTECH_DRAGON_KILL, false],
        [DragonType.HEXTECH, 'Hextech', EventType.ENEMY_HEXTECH_DRAGON_KILL, false],
        [DragonType.ELDER, 'Elder', EventType.ENEMY_ELDER_DRAGON_KILL, false],
    ];

    test.each(dragonToExpectedTypeArguments)(
        '%s -> %s (allied=%s)',
        (_dragonType, apiType, expectedEventType, alliedKill) => {
            //given
            setupGameState(alliedKill);
            const event = dragonEvent(apiType);

            //when
            const eventType = eventTypeFromEvent(event);

            //then
            expect(eventType).toBe(expectedEventType);
        },
    );

    test('testActivePlayerKill', () => {
        //given
        setupGameState(true);
        const event: Event = { EventID: 14, EventName: 'ChampionKill', EventTime: 1118, KillerName: PLAYER_RIOT_ID_GAME_NAME, VictimName: 'Warwick Bot', Assisters: [] };

        //when
        const eventType = eventTypeFromEvent(event);

        //then
        expect(eventType).toBe(EventType.ACTIVE_PLAYER_KILL);
    });

    test('testActivePlayerAssist', () => {
        //given
        setupGameState(true);
        const event: Event = { EventID: 14, EventName: 'ChampionKill', EventTime: 1118, KillerName: 'Some Champ', VictimName: 'Warwick Bot', Assisters: [PLAYER_RIOT_ID_GAME_NAME] };

        //when
        const eventType = eventTypeFromEvent(event);

        //then
        expect(eventType).toBe(EventType.ACTIVE_PLAYER_ASSIST);
    });

    test('testAllyHeraldKillEvent', () => {
        //given
        setupGameState(true);
        const event: Event = { EventID: 7, EventName: 'HeraldKill', EventTime: 933, KillerName: 'BooonE' };

        //when
        const eventType = eventTypeFromEvent(event);

        //then
        expect(eventType).toBe(EventType.ALLY_HERALD_KILL);
    });

    test('testEnemyHeraldKillEvent', () => {
        //given
        setupGameState(false);
        const event: Event = { EventID: 7, EventName: 'HeraldKill', EventTime: 933, KillerName: 'BooonE' };

        //when
        const eventType = eventTypeFromEvent(event);

        //then
        expect(eventType).toBe(EventType.ENEMY_HERALD_KILL);
    });

    test('testAllyBaronKillEvent', () => {
        //given
        setupGameState(true);
        const event: Event = { EventID: 7, EventName: 'BaronKill', EventTime: 933, KillerName: 'BooonE' };

        //when
        const eventType = eventTypeFromEvent(event);

        //then
        expect(eventType).toBe(EventType.ALLY_BARON_KILL);
    });

    test('testEnemyBaronKillEvent', () => {
        //given
        setupGameState(false);
        const event: Event = { EventID: 7, EventName: 'BaronKill', EventTime: 933, KillerName: 'BooonE' };

        //when
        const eventType = eventTypeFromEvent(event);

        //then
        expect(eventType).toBe(EventType.ENEMY_BARON_KILL);
    });

    test('testGameEndWinEvent', () => {
        //given
        setupGameState(true);
        const event: Event = { EventID: 26, EventName: 'GameEnd', EventTime: 1375, Result: 'Win' };

        //when
        const eventType = eventTypeFromEvent(event);

        //then
        expect(eventType).toBe(EventType.GAME_END_VICTORY);
    });

    test('testGameEndLoseEvent', () => {
        //given
        setupGameState(true);
        const event: Event = { EventID: 26, EventName: 'GameEnd', EventTime: 1375, Result: 'Lose' };

        //when
        const eventType = eventTypeFromEvent(event);

        //then
        expect(eventType).toBe(EventType.GAME_END_DEFEAT);
    });
});
